app/src/main/java/com/nextcloud/talk/jobs/UploadAndShareFilesWorker.kt

```diff
@@ -0,0 +1,184 @@
+import com.nextcloud.talk.models.database.UserEntity
+import com.nextcloud.talk.models.json.generic.GenericOverall
+import com.nextcloud.talk.utils.ApiUtils
+import com.nextcloud.talk.utils.bundle.BundleKeys.KEY_FILE_PATHS
+import com.nextcloud.talk.utils.bundle.BundleKeys.KEY_INTERNAL_USER_ID
+import com.nextcloud.talk.utils.bundle.BundleKeys.KEY_ROOM_TOKEN
+import com.nextcloud.talk.utils.database.user.UserUtils
+import com.nextcloud.talk.utils.preferences.AppPreferences
+import io.reactivex.Observer
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
+import okhttp3.MediaType
+import okhttp3.RequestBody
+import retrofit2.Response
+import java.io.FileNotFoundException
+import java.io.IOException
+import java.io.InputStream
+import java.util.*
+import javax.inject.Inject
+
+
+@AutoInjector(NextcloudTalkApplication::class)
+class UploadAndShareFilesWorker(val context: Context, workerParameters: WorkerParameters) :
+        Worker(context, workerParameters) {
+
+    @Inject
+    lateinit var ncApi: NcApi
+
+    @Inject
+    lateinit var userUtils: UserUtils
+
+    @Inject
+    lateinit var appPreferences: AppPreferences
+
+    override fun doWork(): Result {
+        NextcloudTalkApplication.sharedApplication!!.componentApplication.inject(this)
+
+        try {
+            val currentUser = userUtils.currentUser
+            val sourcefiles = inputData.getStringArray(DEVICE_SOURCEFILES)
+            val ncTargetpath = inputData.getString(NC_TARGETPATH)
+            val roomToken = inputData.getString(ROOM_TOKEN)
+
+            checkNotNull(currentUser)
+            checkNotNull(sourcefiles)
+            require(sourcefiles.isNotEmpty())
+            checkNotNull(ncTargetpath)
+            checkNotNull(roomToken)
+
+            for (index in sourcefiles.indices) {
+                val sourcefileUri = Uri.parse(sourcefiles[index])
+                var filename = getFileName(sourcefileUri)
+                val requestBody = createRequestBody(sourcefileUri)
+                uploadFile(currentUser, ncTargetpath, filename, roomToken, requestBody)
+            }
+        } catch (e: IllegalStateException) {
+            Log.e(javaClass.simpleName, "Something went wrong when trying to upload file", e)
+            return Result.failure()
+        } catch (e: IllegalArgumentException){
+            Log.e(javaClass.simpleName, "Something went wrong when trying to upload file", e)
+            return Result.failure()
+        }
+        return Result.success()
+    }
+
+    private fun createRequestBody(sourcefileUri: Uri): RequestBody? {
+        var requestBody: RequestBody? = null
+        try {
+            val input: InputStream = context.contentResolver.openInputStream(sourcefileUri)!!
+            val buf: ByteArray
+            buf = ByteArray(input.available())
+            while (input.read(buf) != -1);
```

Comment:
This is an empty while loop?

Reply:
yes. see https://mkyong.com/java/how-to-read-file-in-java-fileinputstream/#fileinputstream---better-performance